File: test/fakers/sku_faker.dart
import 'package:faker/faker.dart';
import 'package:sertton/core/catalog/dtos/sku_dto.dart';
import 'package:sertton/core/catalog/dtos/variation_dto.dart';
import 'variation_faker.dart';

class SkuFaker {
  static final _faker = Faker();

  static SkuDto fakeDto({
    String? id,
    String? skuCode,
    double? costPrice,
    double? salePrice,
    double? discountPrice,
    double? weight,
    double? height,
    double? width,
    double? length,
    String? imageUrl,
    List<VariationDto>? variations,
    int? stock,
    String? yampiToken,
  }) {
    final effectiveSalePrice =
        salePrice ?? _faker.randomGenerator.decimal(scale: 100, min: 10);
    final effectiveDiscountPrice = discountPrice ?? effectiveSalePrice * 0.9;
    final effectiveCostPrice = costPrice ?? effectiveSalePrice * 0.6;

    return SkuDto(
      id: id ?? _faker.guid.guid(),
      skuCode:
          skuCode ??
          'SKU-${_faker.randomGenerator.integer(999999, min: 100000)}',
      costPrice: effectiveCostPrice,
      salePrice: effectiveSalePrice,
      discountPrice: effectiveDiscountPrice,
      weight: weight ?? _faker.randomGenerator.decimal(scale: 5, min: 0.1),
      height: height ?? _faker.randomGenerator.decimal(scale: 50, min: 1),
      width: width ?? _faker.randomGenerator.decimal(scale: 50, min: 1),
      length: length ?? _faker.randomGenerator.decimal(scale: 50, min: 1),
      imageUrl: imageUrl ?? _faker.image.loremPicsum(),
      variations: variations ?? VariationFaker.fakeManyDto(count: 2),
      stock: stock ?? _faker.randomGenerator.integer(100, min: 0),
      yampiToken: yampiToken ?? _faker.guid.guid(),
    );
  }

  static List<SkuDto> fakeManyDto({
    int count = 10,
    String? id,
    String? skuCode,
    double? costPrice,
    double? salePrice,
    double? discountPrice,
    double? weight,
    double? height,
    double? width,
    double? length,
    String? imageUrl,
    List<VariationDto>? variations,
    int? stock,
    String? yampiToken,
  }) {
    return List.generate(
      count,
      (index) => fakeDto(
        id: id,
        skuCode: skuCode,
        costPrice: costPrice,
        salePrice: salePrice,
        discountPrice: discountPrice,
        weight: weight,
        height: height,
        width: width,
        length: length,
        imageUrl: imageUrl,
        variations: variations,
        stock: stock,
        yampiToken: yampiToken,
      ),
    );
  }
}

File: src/core/catalog/entities/fakers/sku-faker.ts
import { faker } from "@faker-js/faker"

import type { Sku } from "../sku"

export class SkuFaker {
  static fake(sku?: Partial<Sku>): Sku {
    const salePrice = faker.number.float({ fractionDigits: 2, min: 10, max: 1000 })

    return {
      costPrice: salePrice * 0.6,
      discountPrice: salePrice * 0.9,
      height: faker.number.float({ fractionDigits: 2, min: 1, max: 100 }),
      id: faker.string.uuid(),
      imageUrl: faker.image.url(),
      length: faker.number.float({ fractionDigits: 2, min: 1, max: 100 }),
      salePrice,
      skuCode: faker.string.alphanumeric(8).toUpperCase(),
      stock: faker.number.int({ min: 1, max: 100 }),
      variations: [],
      weight: faker.number.float({ fractionDigits: 2, min: 0.1, max: 50 }),
      width: faker.number.float({ fractionDigits: 2, min: 1, max: 100 }),
      yampiToken: faker.string.alphanumeric(20),
      ...sku,
    }
  }

  static fakeMany(count = 10, sku?: Partial<Sku>): Sku[] {
    return Array.from({ length: count }, () => SkuFaker.fake(sku))
  }
}
